libraries/Microsoft.Bot.Builder.Azure/CosmosDbSqlStorage.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.Documents.Linq;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// Models IStorage around a CosmosDB SQL (DocumentDB)
+    /// </summary>
+    public class CosmosDbSqlStorage : IStorage
+    {
+        private readonly string databaseId;
+        private readonly string collectionId;
+        private readonly DocumentClient client;
+
+        private const string UserAgentSuffix = "Microsoft-BotFramework 4.0.0";
+
+        private static JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All
+        });
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CosmosDbSqlStorage"/> class,
+        /// using the provided CosmosDB credentials, DatabaseId and CollectionId.
+        /// </summary>
+        /// <param name="serviceEndpoint">The endpoint Uri for the service endpoint from the Azure Cosmos DB service.</param>
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        public CosmosDbSqlStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            this.databaseId = databaseId;
```

Comment:
our coding standard has been to (generally) use "_" for member variable names and omit the "this.".

Reply:
Done!